File: python/0387_first_unique_character_in_a_string.py
# n = len(s)
# Time: O(n)
# Space: O(n)
from collections import defaultdict


class Solution:
    def firstUniqChar(self, s: str) -> int:
        count_map = defaultdict(int)

        for char in s:
            count_map[char] = count_map.get(char, 0) + 1

        for idx, char in enumerate(s):
            if count_map[char] == 1:
                return idx

        return -1

File: go/0387_first_unique_character_in_a_string.go
package main

// n = len(s)
// Time:  O(n)
// Space: O(n)
func firstUniqChar(s string) int {
	freq := make(map[rune]int)
	for _, char := range s {
		freq[char]++
	}

	for idx, char := range s {
		if freq[char] == 1 {
			return idx
		}
	}

	return -1
}
